### ISSUE TITLE
twisted.trial.unittest.TestCase.assertWarns always fails when the C warnings module, new in Python trunk, is in use

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3223]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3223|
|Type|defect|
|Created|2008-05-06 13:52:41Z|
|Branch|https://github.com/twisted/twisted/tree/catch_warnings-3223|

`assertWarns` relies on being able to replace `warnings.warn_explicit` to learn when warnings are issued.  It seems this won't work in Python 2.6, as the warning module is now partially implemented in C and ignores the `warn_explicit` attribute of the `warnings` module, instead calling its own internal version.

<details><summary>Searchable metadata</summary>

```
trac-id__3223 3223
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__Python_2_6 Python-2.6
branch__branches_catch_warnings_3223 branches/catch_warnings-3223
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1210081961000000 1210081961000000
changetime__1221501828000000 1221501828000000
version__None None
owner__ 
cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

After a quick discussion with exarkun, we conclude that the way to go is to use callDeprecated and deprecate assertWarns (sigh). Maintaining something like that is too much a burden anyway. So I'd say the only thing to do in this ticket is deprecate assertWarns, and open tickets for its usage in twisted itself.

(BTW jml, thank you for not listening to me and not using warn_explicit).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

It looks like there will be a way to do this in Python 2.6 after all.  `warnings.catch_warnings` will keep track of filters, so a simplefilter of "always" can be installed with it to avoid duplicate suppression.  Let me rephrase that in Python to be more clear:

```python
    if sys.version_info >= (2, 6):
        def assertWarns(self, category, message, filename, f, *args, **kwargs):
            with warnings.catch_warnings(True) as w:
                warnings.simplefilter("always")
                result = f(*args, **kwargs)
                # test category, message, lineno, filename against attributes of w
                return result
    else:
        # The old thing
```

Of course, something to consider here is that we can't actually use the `with` syntax, so we may have to emulate it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

OK I feel terrible of suggesting that, but what about using exec?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Might be worthwhile in this case, since there's very little code (we actually only need the first two lines to be inside the `with` suite).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [24722]) Branching to 'catch_warnings-3223'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

This seems to be the right thing now.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @exarkun|
|-|-|

Great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [24853]) Merge catch_warnings-3223

Author: exarkun
Reviewer: therve
Fixes: [#3223](https://github.com/twisted/twisted/issues/3223)

Add support to `TestCase.failUnlessWarns` for using the new
`warnings.catch_warnings` API in Python 2.6.  This ensures
that warnings will actually be observed by `failUnlessWarns`
in newer versions of Python where the old strategy, which
relied on overriding `warnings.warn_explicit`, can no longer
be relied upon.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
